bugsnag-android-core/src/main/java/com/bugsnag/android/Journal.java

```diff
@@ -0,0 +1,170 @@
+package com.bugsnag.android;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.dslplatform.json.DslJson;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A journal represents a series of commands that will manipulate a document.
+ * Sometimes it's more efficient to store a base document and a list of modifications
+ * to me made to it rather than re-serialize the entire document after every modification.
+ */
+public class Journal {
+    @NonNull
+    public final List<Command> commands;
+
+    public Journal() {
+        commands = new LinkedList<>();
+    }
+
+    /**
+     * Apply this journal to a document.
+     *
+     * This method goes through the journal entries one-by-one, applying their modification
+     * commands to the document. This is a destructive operation that will modify the document.
+     *
+     * Note: If a journal command replaces the root object, the resulting document will be a
+     * different object from the one passed in.
+     *
+     * @param document The document to apply to.
+     * @return The finished document (may be a different object from the one passed in).
+     */
+    public @NonNull Object applyTo(@NonNull Object document) {
```

Comment:
I'm not clear what type is expected to be passed in and return here. All the examples use `Map`, which would be preferable to `Object` as it's more precise. Unless there are good reasons not to, I think these methods need to use types, as it gives type safety and will presumably reduce the number of expensive type checks by callers of this API.